### PR TITLE
Dashboard tests

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Implements https://github.com/enso-org/cloud-v2/issues/631
  - Tests for dashboard (`app/ide-desktop/lib/dashboard/`):
    - End-to-end tests
    - Unit tests
    - Component tests

The purpose of this PR is to introduce the testing framework - more tests can be added later in separate PRs.

### Important Notes
To test, run `npm run test` in `app/ide-desktop`, or `app/ide-desktop/lib/dashboard/`. All tests should pass.

Individual test types can be run using `npm run test-unit`, `npm run test-component` and `npm run test-e2e` in `app/ide-desktop/lib/dashboard/`.
Individual end-to-end tests can be run using `npx playwright test -c playwright-e2e.config.ts test-e2e/<file name>.spec.ts` in `app/ide-desktop/lib/dashboard/`.

End-to-end tests require internet access to pass (for things like fonts).

This PR *does* check in screenshots to guard against visual regessions (and/or to make visual changes obvious)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

**Comment by PabloBuchu** [comment]
Lots of failed tests with missing snapshot errror.. 

Screenshot 2023-08-25 at 10 00 25

**Comment by somebody1234** [comment]
@PabloBuchu that seems to be because the snapshots are platform-specific. i think that's a good thing, because some things (like context menus) *are* different on each platform. i *think* the correct move here would be to `git add -A` + commit + push?

**Comment by somebody1234** [comment]
speaking of though, that means these should be run on windows too

**Comment by somebody1234** [comment]
lint check is failing though, one moment...

**Comment by somebody1234** [comment]
I think this *should* have some CI changes, since we *should* test for visual regressions on all platforms, but I think maybe that should wait until it's confirmed that we want to merge this (or wait until this is actually merged in, so that there's something to run in those CI actions)

**Comment by somebody1234** [comment]
for anyone testing: this does have new folders (caches, test results) which are `.gitignore`d, remember to `git add -A; git reset --hard` when switching away from this branch

**Review by PabloBuchu — APPROVED** [review]

**Comment by somebody1234** [comment]
notes:
- github stats says a lot of lines have been changed, but i think almost all of it is due to a lockfile version change
- screenshots probably (?) need to be regenerated on the other platforms again.

- CI is *running* the E2E tests at least, however:
  - it should probably cache the browser binaries
  - it is consistently giving *very* different results (50%-85% difference) compared to my pc (artix linux)... even though it's supposed to be headless, meaning that distro shouldn't matter. some possible explanations:
    - CI is potentially not loading the fonts
    - CI is not loading the background?
    - CI not using the GPU shouldn't be the issue, as the tests are explicitly using `swiftshader` as the ANGLE backend.

**Comment by somebody1234** [comment]
i think it's better to wait until michal gets back so we can get the failing screenshots from the CI runner and figure out what is going wrong

**Comment by somebody1234** [comment]
@mwu-tow would it be possible for you to get the images out of the failing test run? (you might have to re-run CI because the test results have probably been cleared from all runners)

(i can't reproduce the failing tests locally, even though i am on linux)

File: app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts]: Looks like a lot of duplicate code here (and throughout the rest of the file) from the non-mock definition of the class. If this is a mock, shouldn't it re-use the same types/interfaces and only change the underlying `impl`s?

File: app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts]: What's this `aJwtToken.` bit? A mock JWT header? Should probably put it in a constant with some docs.

File: app/ide-desktop/lib/dashboard/package.json
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/package.json]: `FIXME[sb]` -> `FIXME [sb]`

Comment (indiv0) [reply app/ide-desktop/lib/dashboard/package.json]: Looks like it's part of the `test` script again so this comment can be removed?

File: app/ide-desktop/lib/dashboard/test-e2e/actions.ts
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/test-e2e/actions.ts]: `regresison` -> `regression`

**Review by indiv0 — CHANGES_REQUESTED** [review]

File: app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts]: for now i've just removed the header instead, doesn't seem like it's needed

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts:393]: for now i've just removed the header instead, doesn't seem like it's needed

File: app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts]: good point. i didn't think of that due to how the mock files are injected (?) into the app, but it *is* a simple change to import things from the original files

File: app/ide-desktop/lib/dashboard/package.json
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/package.json]: yeah. the other comments can too, since i switched from headful (because headless wasn't properly rendering `backdrop-filter: blur`) back to headless

**Review by indiv0 — APPROVED** [review]

**Comment by PabloBuchu** [comment]
@somebody1234 lets fix conflitcs and merge it 
are we able to do something about https://github.com/enso-org/enso/actions/runs/6161014611/job/17005311659?pr=7656?

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Dashboard tests · enso-org/enso@510af5f</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by somebody1234** [comment]
@PabloBuchu yeah i think we still can't merge because lint is failing, because the screenshots are different between ci and desktop. kinda forgot about this, i'll merge and then ask michal if something can be done about the lint screenshot files

**Comment by somebody1234** [comment]
://// `npm run test` is hanging on my pc... no amount of debugging so far has given so much as a lead

**Comment by somebody1234** [comment]
what, why did lint hang

**Comment by somebody1234** [comment]
:D lint finally passes!!!

**Comment by somebody1234** [comment]
this *should* still get updated screenshots from macos and windows though. (cc @PabloBuchu)

**Comment by somebody1234** [comment]
:D (visual regression tests)

image

**Pull request merged**
